### ISSUE TITLE
StrictMode: getDefaultSharedPreferences()

### ISSUE THREAD

**Issue by grote** [mikepenz/MaterialDrawer#2238]
## About this issue

When investigating a UI hang in my app, I noticed that MaterialDrawer is violating StrictMode by [calling getDefaultSharedPreferences()](
https://github.com/mikepenz/MaterialDrawer/blob/b903da90c60ae46cc122a3e319112098fa596141/library/src/main/java/com/mikepenz/materialdrawer/DrawerBuilder.java#L1254) on the `UiThread`.

In my case, this hangs for quite some time on a Nexus 5X:

```
D/StrictMode: StrictMode policy violation; ~duration=1603 ms: android.os.StrictMode$StrictModeDiskReadViolation: policy=65599 violation=2
                                     at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1440)
                                     at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:251)
                                     at java.io.File.exists(File.java:807)
                                     at android.app.ContextImpl.getDataDir(ContextImpl.java:2197)
                                     at android.app.ContextImpl.getPreferencesDir(ContextImpl.java:517)
                                     at android.app.ContextImpl.getSharedPreferencesPath(ContextImpl.java:714)
                                     at android.app.ContextImpl.getSharedPreferences(ContextImpl.java:368)
                                     at android.content.ContextWrapper.getSharedPreferences(ContextWrapper.java:167)
                                     at android.preference.PreferenceManager.getDefaultSharedPreferences(PreferenceManager.java:526)
                                     at com.mikepenz.materialdrawer.DrawerBuilder.handleShowOnLaunch(DrawerBuilder.java:1252)
                                     at com.mikepenz.materialdrawer.DrawerBuilder.buildView(DrawerBuilder.java:1550)
                                     at com.mikepenz.materialdrawer.DrawerBuilder.build(DrawerBuilder.java:1333)
```

A possible solution could be to allow the user to pass in the `SharedPreferences` that the drawer then uses instead of retrieving them again from disk.

## Details

- Used library version: 6.0.0
- Used support library version: 27.0.2
- Used gradle build tools version: 27.0.3
- Used tooling / Android Studio version: 3.0.1
- Other used libraries, potential conflicting libraries

## Checklist

- [x] Searched for [similar issues](https://github.com/mikepenz/MaterialDrawer/issues)
- [x] Checked out the [sample application](https://github.com/mikepenz/MaterialDrawer/tree/develop/app)
- [x] Read the [README](https://github.com/mikepenz/MaterialDrawer/blob/develop/README.md)
- [x] Checked out the [CHANGELOG](https://github.com/mikepenz/MaterialDrawer/releases)
- [x] Read the [FAQ](https://github.com/mikepenz/MaterialDrawer/blob/develop/FAQ.md)
- [x] Checked out the [MIGRATION GUIDE](https://github.com/mikepenz/MaterialDrawer/blob/develop/MIGRATION.md)

**Comment by EfremovAV** [comment]
How long to wait for enhancement?

**Comment by EfremovAV** [comment]
Does the problem depend on the OS version or device model?

**Comment by mikepenz** [comment]
@grote that sounds really strange. why would a simple preference read be so slow on that phone. 

I would be super happy about a PR to enhance this. For example it could be that the preferences are only read in if the 2 features are activated. 

In addition perhaps it should not use the default preferences (as they might be super big based on the app). But this might be handled as you mentioned, by allowing to pass in the preferences.

**Comment by grote** [comment]
I agree it is very strange. Normally, getting the preferences should be only a few ms, but still I had the above stack trace in my logcat where it took over a second. Either there's a bug somewhere further down the stack, or this is because during application startup, there's so much disk IO that it gets congested.

I'd be happy to provide a PR, but need to see how I can test custom version of your library. I see there's an example app bundled with it, so I'll probably use that one.

**Comment by EfremovAV** [comment]
This [sample ](https://github.com/tral/MaterialDrawerSample) works fine on the Nexus 5x. But there uses the old version of the library 0.9.5.

**Comment by mikepenz** [comment]
**6.0.7-SNAPSHOT** @grote created a snapshot release which will not load the preferences if the 2 features which need them are disabled:
https://github.com/mikepenz/MaterialDrawer/commit/f6d181e08ff49603a35f25b90f57ac3d8fd7d392

can you please test if it makes a difference?

**Comment by grote** [comment]
@mikepenz thanks, that's great news! Is this SNAPSHOT release available in any repository or you do you suggest testing this?

**Comment by mikepenz** [comment]
you need to use the `SNAPSHOT` maven central repo. It's just so we can make sure not loading the shared preferences speeds it up. (you have to disable those 2 functionalities though)

**Comment by grote** [comment]
No luck with `mavenCentral()` either: http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22com.mikepenz%22%20AND%20a%3A%22materialdrawer%22

But it works with jitpack.io: `implementation 'com.github.mikepenz:MaterialDrawer:develop-SNAPSHOT'`

I am not using these 2 features that require the preferences anyway. So without those, I tested the new version and it is not causing strict mode violations anymore and speeds things up.

> When investigating a UI hang in my app, I noticed that MaterialDrawer is violating StrictMode

It had turned out that this was unrelated to MaterialDrawer (in my case). But only loading the preferences from disk when needed is a good improvement which is nice to have in any case. Thanks for making it!

**Comment by mikepenz** [comment]
https://github.com/mikepenz/MaterialDrawer/blob/develop/build.gradle#L31

**Comment by EfremovAV** [comment]
My problem has disappeared. I don't know what it is connected with, but there were 3 events:
1. Android Studio has been updated to the latest version.
2. The phone has installed an update.
3. I reduced the size of the logo in the header.

**Comment by mikepenz** [comment]
@EfremovAV ok in your case it most likely was the logo :D

**Comment by grote** [comment]
Since the commit that addresses the problem landed, we could close this issue now. Thanks Mike!

**Comment by EfremovAV** [comment]
Maybe, but then why was there no problem on other phones?

**Comment by mikepenz** [comment]
@grote the commit only solves it if you disable these features, I'd love to check if it solves it before allowing to pass in the shared preferences or have an alternative. 
Had you time to run it once?

@EfremovAV perhaps the phone was just low on memory or so. there are many factors which could lead to some strange behaviour

**Comment by grote** [comment]
> Had you time to run it once?

Yes, as I wrote above:

> I am not using these 2 features that require the preferences anyway. So without those, I tested the new version and it is not causing strict mode violations anymore and speeds things up.

Having the option to pass in the preferences would of course put the cherry on top of things ;)

**Comment by mikepenz** [comment]
Added the option to pass in the shared preferences. released in v6.0.7 

https://github.com/mikepenz/MaterialDrawer/commit/4d7b0d41621e42a2b68506ec1b48d2c1bbb2ce56
